pkg/controller/clusterapimachine/machine_controller.go

```diff
@@ -0,0 +1,300 @@
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(ms *clusterv1.Machine) {
+	key, err := controller.KeyFunc(ms)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", ms, err))
+		return
+	}
+
+	c.queue.Add(key)
+}
+
+// enqueueAfter will enqueue a cluster after the provided amount of time.
+func (c *Controller) enqueueAfter(cluster *clustopv1.Cluster, after time.Duration) {
+	key, err := controller.KeyFunc(cluster)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for object %#v: %v", cluster, err))
+		return
+	}
+
+	c.queue.AddAfter(key, after)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Sync %q failed with %v", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+// syncMachine will sync the machine with the given key.
+// This function is not meant to be invoked concurrently with the same key.
+func (c *Controller) syncMachine(key string) error {
+	startTime := time.Now()
+	c.logger.WithField("key", key).Debug("syncing machine")
+	defer func() {
+		c.logger.WithFields(log.Fields{
+			"key":      key,
+			"duration": time.Now().Sub(startTime),
+		}).Debug("finished syncing machine")
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	machine, err := c.machinesLister.Machines(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.logger.WithField("key", key).Debug("machine has been deleted")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	mLog := colog.WithClusterAPIMachine(c.logger, machine)
+	mLog.Debugf("found machine")
+
+	// Lookup the cluster for this machine:
+	clusterName, ok := machine.Labels[clusterNameLabel]
+	if !ok {
+		return fmt.Errorf("no cluster label set on machine")
+	}
+	cluster, err := c.clustersLister.Clusters(namespace).Get(clusterName)
+	if errors.IsNotFound(err) {
+		mLog.WithField("cluster", clusterName).Errorf("no cluster found for machine")
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	clusterSpec, err := controller.ClusterSpecFromClusterAPI(cluster)
+	if err != nil {
+		return err
+	}
+
+	// TODO: once cluster controller is setting resolved cluster version refs on the
+	// ClusterStatus, replace this with the resolved ref not a lookup
+	clusterVersion, err := c.clustopClient.ClusteroperatorV1alpha1().ClusterVersions(clusterSpec.ClusterVersionRef.Namespace).Get(clusterSpec.ClusterVersionRef.Name, metav1.GetOptions{})
```

Comment:
Why do we use a lister to fetch the Cluster but the client directly to fetch the ClusterVersion?

Reply:
Which should be used when?

Reply:
I don't know the official stance on this. My ruke of thumb is to prefer Lister for any type that is used regularly and of which we need a large subset of the total population. ClusterVersion fits into that description. Lister saves us from having to go back to the API Server to service the query at the cost of having to store the objects locally and watch the API Server for changes.